### PR TITLE
feat(qwen): add tensor parallel support for Qwen3-0.6B/4B

### PR DESCRIPTION
**Pull request by daichu-nv**
## Background
Qwen needs TensorRT 11 tensor-parallel build and runtime support without changing the normal single-device Qwen path. This PR keeps the original Qwen builder route intact and adds TP=4 E2E coverage for both Qwen3-0.6B and Qwen3-4B-Instruct-2507.

The latest premerge failure was in selective E2E, not the TP implementation itself. PR impact analysis emitted `ci_tier: multi_device` manifests into `e2e_models.txt`, but the CI runner path is still single-device and invokes E2E collection without `--multi-device-only`. That produced pytest collection failures on the TP4 node IDs.

## Exit Criteria
- Single-device Qwen continues to use the existing `standard_decoder_builder.py` / `dual_profile_decoder_builder.py` path.
- Qwen TP builds package one rank-local engine per rank and run through the C++ runtime with `mpirun -np <tp_size>`.
- TP=4 accuracy is validated locally on NVIDIA B300 x4 for both Qwen3-0.6B and Qwen3-4B-Instruct-2507.
- PR/L0 CI and nightly CI do not run multi-device E2E until CI has a runner pool that can reserve all GPUs for tensor-parallel cases.
- Manual/local multi-device E2E remains available through explicit opt-in flags.

## Implementation
- Added `tensorrt_model_connect/tensorrt_model_connect/parallel_config.py` to normalize TP config, validate supported sizes, name rank-local bundle sections, and shard standard decoder weights.
- Added `tensorrt_model_connect/tensorrt_model_connect/families/qwen/dual_profile_decoder_tp_builder.py` as the TP companion to the existing dual-profile builder. It intentionally duplicates the non-TP graph where behavior is the same, but makes TP-specific sharding, rank-local head/KV dimensions, and `ALL_REDUCE` joins explicit.
- Updated `tensorrt_model_connect/tensorrt_model_connect/families/qwen/plugin.py` so single-device Qwen still calls `build_standard_decoder_engine`, while TP calls the new dual-profile TP builder.
- Added TP=4 multi-device E2E manifests for `qwen3-0.6b-fp16-tp4` and `qwen3-4b-instruct-2507-tp4`, using default text-generation thresholds so generated output and logit parity are both gated. These manifests are tagged with `ci_tier: "multi_device"`.
- Added distributed runtime wiring in C++ for TP bundles: rank-local engine loading, NCCL communicator setup, and distributed context lifetime ordering.
- Extended the E2E harness so multi-device cases are opt-in with `--multi-device-only`, launch TRT and debug-logit collection with `mpirun`, and compare full per-step logits against HF.
- Added the reviewed test-impact allowlist entry for `parallel_config.py`, which fixes the earlier impact-analysis failure for this new shared builder module.
- Updated impact selection so `tools/test_impact.py` excludes `ci_tier: multi_device` by default. Manual/local analysis can opt back in with `--include-ci-tier multi_device`.
- Updated nightly full E2E separately because it bypasses impact analysis. `run_full_e2e` now passes `--exclude-ci-tier multi_device` to both `scripts/warm_hf_cache.py` and `scripts/run_e2e_parallel.sh`.
- Added TODOs at both CI exclusion sites to remove the `multi_device` exclusions once CI has a runner pool that can reserve all GPUs for tensor-parallel E2E.

There are two CI paths to keep in mind:

1. PR/L0 selective CI uses `tools/test_impact.py`. The default exclusion keeps `impact.json` and `e2e_models.txt` single-device because CI does not pass `--include-ci-tier multi_device`.
2. Nightly full E2E does not use `tools/test_impact.py` to choose models. It directly warms and collects all manifests, so `.github/scripts/run-trtmc-ci.sh` must explicitly pass `--exclude-ci-tier multi_device` on that path.

## Manual Multi-Device Runs
To include multi-device models in local impact analysis:

```bash
python3 tools/test_impact.py   --files tests/e2e/models/qwen3-0.6b-fp16-tp4.json   --include-ci-tier multi_device   --json
```

To run the Qwen3-0.6B TP4 E2E manually on a 4-GPU machine:

```bash
python3 -m pytest 'tests/test_e2e.py::test_e2e[qwen3-0.6b-fp16-tp4]' -vv -rs   --multi-device-only   --rebuild-engines   --engine-dir /tmp/trtmc-e2e-qwen06b-tp4/engines   --e2e-artifacts-dir /tmp/trtmc-e2e-qwen06b-tp4/artifacts   --trtmc-binary build/trtmc   --hf-python /opt/venv/bin/python3
```

To run the Qwen3-4B-Instruct-2507 TP4 E2E manually on a 4-GPU machine:

```bash
python3 -m pytest 'tests/test_e2e.py::test_e2e[qwen3-4b-instruct-2507-tp4]' -vv -rs   --multi-device-only   --rebuild-engines   --engine-dir /tmp/trtmc-e2e-qwen4b-tp4/engines   --e2e-artifacts-dir /tmp/trtmc-e2e-qwen4b-tp4/artifacts   --trtmc-binary build/trtmc   --hf-python /opt/venv/bin/python3
```

## Validation
- `python3 -m pytest tests/builder/test_parallel_config.py tests/builder/test_family_plugins.py tests/builder/test_debug_runner.py -q -p no:cacheprovider`: passed, 71 tests.
- `python3 tools/test_impact.py --validate`: passed; existing warnings only.
- `python3 -m pytest tests/tools/test_test_impact.py -q -p no:cacheprovider`: passed, 124 tests.
- `.venv-pytest/bin/python -m pytest tests/tools/test_test_impact.py -q`: passed, 126 tests.
- `.venv-pytest/bin/python -m pytest tests/tools/test_github_actions_ci.py tests/tools/test_warm_hf_cache_static.py -q`: passed, 16 tests.
- `bash -n .github/scripts/run-trtmc-ci.sh`: passed.
- `python3 -m py_compile tools/test_impact.py`: passed.
- `python3 tools/test_impact.py --files tests/e2e/models/qwen3-0.6b-fp16-tp4.json --json`: passed; default output has no E2E models because `multi_device` is excluded.
- `python3 tools/test_impact.py --files tests/e2e/models/qwen3-0.6b-fp16-tp4.json --include-ci-tier multi_device --json`: passed; output includes `qwen3-0.6b-fp16-tp4`.
- `python3 -m pytest tests/test_e2e.py::test_e2e[qwen3-0.6b-fp16-tp4] -vv -rs --multi-device-only --rebuild-engines --engine-dir /tmp/trtmc-e2e-qwen06b-tp4/engines --e2e-artifacts-dir /tmp/trtmc-e2e-qwen06b-tp4/artifacts --trtmc-binary build/trtmc --hf-python /opt/venv/bin/python3`: passed on NVIDIA B300 x4, 1 test in 521.62s.
- Qwen3-0.6B TP=4 metrics: `trt_full_logits.npy` and `hf_logits.npy` both had shape `(22, 151936)`, `logit_cosine_p5=0.9999699711799621`, `logit_rel_l2_p95=0.007469473406672476`, `stable_top1_match_rate=1.0`, `token_agreement_rate=1.0`, `normalized_text_edit_distance=0.0`.
- `python3 -m pytest tests/test_e2e.py::test_e2e[qwen3-4b-instruct-2507-tp4] -vv -rs --multi-device-only --rebuild-engines --engine-dir /tmp/trtmc-e2e-qwen4b-tp4/engines --e2e-artifacts-dir /tmp/trtmc-e2e-qwen4b-tp4/artifacts --trtmc-binary build/trtmc --hf-python /opt/venv/bin/python3`: passed on NVIDIA B300 x4, 1 test in 697.09s.
- Qwen3-4B-Instruct-2507 TP=4 metrics: `trt_full_logits.npy` and `hf_logits.npy` both had shape `(22, 151936)`, `logit_cosine_p5=0.9999676167964935`, `logit_rel_l2_p95=0.007996958680450916`, `stable_top1_match_rate=1.0`, `token_agreement_rate=1.0`, `normalized_text_edit_distance=0.0`.

Together, the TP=4 E2E lanes explicitly validate both generated output correctness and full per-step logit parity against the HF reference.

## Notes For Future Readers
- Review the Python TP builder after `parallel_config.py`; the TP builder is deliberately duplicated now to keep the single-device path untouched and the TP graph easy to inspect.
- Future multi-device E2E manifests must set `ci_tier: "multi_device"` so `tests/test_e2e.py --multi-device-only` selects them and normal E2E collection excludes them.
- Do not remove the CI `multi_device` exclusions until CI has a runner pool that can reserve all GPUs for tensor-parallel E2E. The TODOs are intentionally placed next to the default impact exclusion and nightly full-E2E exclusions.

### CODE REVIEW

**Comment by yifeif-nv** [comment]
Block merge for a while. I think we need to discuss should we treat such a TP implementation as an argument or another different blueprint receipt

**Comment by daichu-nv** [comment]
@yifeif-nv  sure, let me create a meeting next Monday, let's sync then, thanks!

File: include/trtmc/runtime/distributed_runtime.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+#include <memory>
+#include <string>
+
+namespace trtmc {
+
+struct DistributedRuntimeGroup {
+    int world_size{1};
+    int rank{0};
+    int local_rank{0};
```
Comment (pkisfaludi-nv) [inline include/trtmc/runtime/distributed_runtime.h]: what is the difference between rank and local_rank?

Comment (daichu-nv) [reply include/trtmc/runtime/distributed_runtime.h]: rank is the global tensor-parallel/MPI rank. It identifies the process inside the TP communicator.

  In this code it is used for:

  - selecting the rank-local engine section: engine_plan_tp_rank{rank}
  - deciding who writes the NCCL rendezvous ID: rank == 0
  - calling ncclCommInitRank(..., tp_size, ..., rank)

  local_rank is the rank on the current host/node, used for GPU binding.

  In this code it is used for:

  - cudaSetDevice(local_rank % device_count)

  Example:

  single node, 4 GPUs:
  rank       = 0 1 2 3
  local_rank = 0 1 2 3

  two nodes, 4 GPUs each:
  rank       = 0 1 2 3 | 4 5 6 7
  local_rank = 0 1 2 3 | 0 1 2 3

So for our current single-node TP4 manual runs, they usually match. But conceptually they are different: rank selects the TP shard/communicator position; local_rank selects the CUDA device visible to that process.

Question: (1) should we only keep one since we focus on single-node first or (2) we add extra comments to clarify things?

Comment (daichu-nv) [reply include/trtmc/runtime/distributed_runtime.h]: I've removed the local_rank in the latest commit. Let's focus on single-node first, thank you!
